### ISSUE TITLE
Feature request: Show a warning if IPFS port is closed

### ISSUE THREAD

**Issue by andrasfuchs** [ipfs-shipyard/ipfs-desktop#1042]
I use ipfs-desktop 0.8.0.0 (and go-ipfs 0.4.21) on Windows 10.

IPFS ports are open on my Windows Firewall and my router by default. If I run IPFS normally I can connect to 250+ peers and my shared files can be accessed without too much extra delay. 

I use a VPN service frequently. This service doesn't support port forwarding, so if I'm on VPN my ports appear to be closed. In this case I can connect to fewer peers (~40) and the access to my files are significantly slower.

I think it would be beneficial to have a small indication somewhere (maybe on the peers page) if the ports on the computer seem to be closed. 
A brief explanation of the drawbacks of this would be motivating, and we could even offer some helpful links to configure the firewall and/or the router.

**Comment by hacdias** [comment]
Thanks for the suggestion @andrasfuchs. I need to see if there's an easy way in Node.js to detect if ports are open on the firewall on all OSes. Do you happen to know if such tool exists?

**Comment by andrasfuchs** [comment]
I don't know Node.js that much, but I'll look into it. Since the checking must be initiated from outside the network of the node in question, it's very likely that we would need to use an external service to do the check. I suppose we would need both ipv4 and ipv6.

As a long term solution it would be great if any node could ask any other node to check its port, and report it back if it found it open or closed ;)

**Comment by hacdias** [comment]
Yes, it should go into the implementation `go-ipfs` and `js-ipfs`. Perhaps you should open an issue there 😃

**Comment by hacdias** [comment]
Just leaving a ref to #1001.

**Comment by hacdias** [comment]
From @andrasfuchs on https://github.com/ipfs/go-ipfs/issues/6593

> I see, alright, thank you @Stebalien for the explanation!
> 
> So, if we wanted to show a warning message to ipfs-desktop users when they are in passive mode (their port looks closed from the outside), we should:
> 1, check if AutoReplay is enabled in their config file (because it's needed for the port checking)
> 2, check if `ipfs id` returns any addresses with the "p2p-circuit" substring (because this means that our port looks closed)
> 
> Is my understanding correct?

**Comment by jessicaschilling** [comment]
@lidel @rafaelramalho19 This sounds like it could be a pretty valid unboarding unblocker - thinking specifically about new users who assume their relevant ports are open, but they're actually not. Is it worth prioritizing?

**Comment by lidel** [comment]
Hm.. dial check would improve onboarding of  new users, but we should avoid using centralized services for the check.

AFAIK go-ipfs ships with [AutoNAT](https://github.com/libp2p/go-libp2p-autonat) enabled to determine if node should be DHT server or only a DHT client:

> an ambient NAT autodiscovery service. It allows peers to figure out their NAT dialability situation by using test dial backs through peers providing the AutoNAT service.

In simpler words, go-ipfs asks publicly dialable peers to check if it is publicly dialable too. 

So in theory, decentralized solution is in place already, but it is internal to go-ipfs and ipfs-desktop can't leverage it to warn user they are not publicly dialabe, because AutoNAT service is not exposed via HTTP API.

@jacobheun  @aschmahmann are you familiar with AutoNAT implementation? 
Got some questions, and would appreciate sanity check on the idea of using it in IPFS Desktop:

- How reliable it is? Is AutoNAT test happening only once, periodically, or on listening address change?
  - Main use case here would be user moving their laptop between different networks: some having UPnP, some blocking IPFS.
- thoughts on exposing the result of AutoNAT test via HTTP API somehow? 
  - for MVP it does not need to be part of official `/api/v0`, it could be under `/debug` like `http://127.0.0.1:5001/debug/metrics/prometheus`
